lib/ddtrace/configuration/components.rb

```diff
@@ -54,28 +54,116 @@ def build_tracer(settings, agent_settings)
+
+          Tracer.new(
             default_service: settings.service,
             enabled: settings.tracer.enabled,
-            partial_flush: settings.tracer.partial_flush.enabled,
+            context_flush: context_flush,
+            sampler: sampler,
+            writer: writer,
             tags: build_tracer_tags(settings),
-            sampler: PrioritySampler.new(
+          )
+        end
+
+        def build_context_flush(settings)
+          if settings.tracer.partial_flush.enabled
+            Datadog::ContextFlush::Partial.new(min_spans_before_partial_flush: settings.tracer.partial_flush.min_spans_threshold)
+          else
+            Datadog::ContextFlush::Finished.new
+          end
+        end
+
+        # TODO: Sampler should be a top-level component.
+        # It is currently part of the Tracer initialization
+        # process, but can take a variety of options (including
+        # a fully custom instance) that makes the Tracer
+        # initialization process complex.
+        def build_sampler(settings)
+          if (sampler = settings.tracer.sampler)
+            if settings.tracer.priority_sampling == false
+              sampler
+            else
+              ensure_priority_sampling(sampler)
+            end
+          elsif settings.tracer.priority_sampling == false
+            Sampling::RuleSampler.new(
+              rate_limit: settings.sampling.rate_limit,
+              default_sample_rate: settings.sampling.default_rate
+            )
+          else
+            PrioritySampler.new(
               base_sampler: AllSampler.new,
               post_sampler: Sampling::RuleSampler.new(
                 rate_limit: settings.sampling.rate_limit,
                 default_sample_rate: settings.sampling.default_rate
               )
             )
+          end
+        end
+
+        def ensure_priority_sampling(sampler = nil)
+          if sampler.is_a?(PrioritySampler)
+            sampler
+          else
+            PrioritySampler.new(
+              base_sampler: sampler,
+              post_sampler: Sampling::RuleSampler.new
+            )
+          end
+        end
+
+        # TODO: Writer should be a top-level component.
+        # It is currently part of the Tracer initialization
+        # process, but can take a variety of options (including
+        # a fully custom instance) that makes the Tracer
+        # initialization process complex.
+        def build_writer(settings, agent_settings)
+          if (writer = settings.tracer.writer)
+            return writer
+          end
+
+          Writer.new(agent_settings: agent_settings, **settings.tracer.writer_options)
+        end
+
+        def subscribe_to_writer_events!(writer, sampler)
+          return unless writer.respond_to?(:events) # Check if it's a custom, external writer
+
+          writer.events.after_send.subscribe(:record_environment_information, &WRITER_RECORD_ENVIRONMENT_INFORMATION_CALLBACK)
```

Comment:
Minor: I would suggest maybe directly inlining this, as I think it ends up being harder to read/follow than a regular block in here.

Reply:
I had this inlined at first, but then I couldn't test what's inside `WRITER_RECORD_ENVIRONMENT_INFORMATION_CALLBACK` and `writer_update_priority_sampler_rates_callback` without making the `subscribe_to_writer_events!` test an integration test. The integration test looked super ugly, with too many options/combinations.